### PR TITLE
[4.0] Improve the cli command discover

### PR DESCRIPTION
**Pull request by ghost**
Pull Request for Issue https://github.com/joomla/joomla-cms/issues/32967 .

### Summary of Changes

I have corrected the misleading duplicate consoles outputs mentioned in #32967 and I have added a task to `install `and a task to `list `the detected extensions.

### Testing Instructions

Run 

`php cli/joomla.php extension:discover:list`

`php cli/joomla.php extension:discover:install `

and 

` php cli/joomla.php extension:discover`

The following is an example:

```
php cli/joomla.php extension:discover:list

 ! [NOTE] Cannot find discovered extensions. Perhaps you need to run extension:discover first?                          

php74:/srv/www/joomla/j4dev# php cli/joomla.php extension:discover:install

 ! [NOTE] There are no discovered extensions for install. Perhaps you need to run extension:discover first?             

php74:/srv/www/joomla/j4dev# php cli/joomla.php extension:discover

 ! [NOTE] 4 extensions have been discovered successfully.                                                               

php74:/srv/www/joomla/j4dev# php cli/joomla.php extension:discover:list

Discovered extensions.
======================

 ------------------------ -------------- --------- ---------- -------- 
  Name                     Extension ID   Version   Type       Active  
 ------------------------ -------------- --------- ---------- -------- 
  en-GB                    306            4.0.0     language   No      
  en-GB                    307            4.0.0     language   No      
  en-GB                    308            4.0.0     language   No      
  plg_sampledata_testing   309            4.0.0     plugin     No      
 ------------------------ -------------- --------- ---------- -------- 

php74:/srv/www/joomla/j4dev# php cli/joomla.php extension:discover:install --eid 306

                                                                                                                        
 [OK] Extension with ID 306 installed successfully.                                                                     
                                                                                                                        

php74:/srv/www/joomla/j4dev# php cli/joomla.php extension:discover:install 

 ! [NOTE] 3 discovered extensions have been installed.     
```

### Actual result BEFORE applying this Pull Request

See https://github.com/joomla/joomla-cms/pull/31524

### Documentation Changes Required

Yes: https://docs.joomla.org/J4.x:CLI_Update

### CODE REVIEW

File: libraries/src/Console/ExtensionDiscoverListCommand.php [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+		$this->configureIO($input, $output);
+
+		$extensions = $this->getExtensions();
+		$state = -1;
+
+		$discovered_extensions = $this->filterExtensionsBasedOnState($extensions, $state);
+
+		if (empty($discovered_extensions))
+		{
+			$this->ioStyle->note("Cannot find discovered extensions. Perhaps you need to run extension:discover first?");
```
Comment (PhilETaylor) [inline libraries/src/Console/ExtensionDiscoverListCommand.php]:
```suggestion
			$this->ioStyle->note("There are no pending discovered extensions to install. Perhaps you need to run extension:discover first?");
```

File: libraries/src/Console/ExtensionDiscoverInstallCommand.php [hunk 2]
```diff
@@ -151,21 +140,71 @@ public function processDiscover($eid): bool
...
+		{
+			return 'Unable to install the extension with ID ' . $eid;
+		}
+		elseif ($count < 0 && $eid < 0)
+		{
+			return 'Unable to install discovered extensions.';
+		}
+		elseif ($count === 0)
+		{
+			return 'There are no discovered extensions for install. Perhaps you need to run extension:discover first?';
```
Comment (PhilETaylor) [inline libraries/src/Console/ExtensionDiscoverInstallCommand.php]:
```suggestion
			return 'There are no pending discovered extensions for install. Perhaps you need to run extension:discover first?';
```

File: libraries/src/Console/ExtensionDiscoverCommand.php [hunk 3]
```diff
@@ -0,0 +1,166 @@
+<?php
+/**
+ * Joomla! Content Management System
+ *
+ * @copyright  (C) 2020 Open Source Matters, Inc. <https://www.joomla.org>
```
Comment (PhilETaylor) [inline libraries/src/Console/ExtensionDiscoverCommand.php]:
```suggestion
 * @copyright  (C) 2021 Open Source Matters, Inc. <https://www.joomla.org>
```

File: libraries/src/Console/ExtensionDiscoverListCommand.php [hunk 4]
```diff
@@ -0,0 +1,109 @@
+<?php
+/**
+ * Joomla! Content Management System
+ *
+ * @copyright  (C) 2020 Open Source Matters, Inc. <https://www.joomla.org>
```
Comment (PhilETaylor) [inline libraries/src/Console/ExtensionDiscoverListCommand.php]:
```suggestion
 * @copyright  (C) 2021 Open Source Matters, Inc. <https://www.joomla.org>
```

File: tests/Unit/Libraries/Cms/Console/ExtensionDiscoverCommandTest.php [hunk 5]
```diff
@@ -0,0 +1,83 @@
+<?php
+/**
+ * @package     Joomla.UnitTest
+ * @subpackage  Console
+ *
+ * @copyright   (C) 2019 Open Source Matters, Inc. <https://www.joomla.org>
```
Comment (PhilETaylor) [inline tests/Unit/Libraries/Cms/Console/ExtensionDiscoverCommandTest.php]:
```suggestion
 * @copyright   (C) 2021 Open Source Matters, Inc. <https://www.joomla.org>
```

File: tests/Unit/Libraries/Cms/Console/ExtensionDiscoverInstallCommandTest.php [hunk 6]
```diff
@@ -0,0 +1,96 @@
+<?php
+/**
+ * @package     Joomla.UnitTest
+ * @subpackage  Console
+ *
+ * @copyright   (C) 2019 Open Source Matters, Inc. <https://www.joomla.org>
```
Comment (PhilETaylor) [inline tests/Unit/Libraries/Cms/Console/ExtensionDiscoverInstallCommandTest.php]:
```suggestion
 * @copyright   (C) 2021 Open Source Matters, Inc. <https://www.joomla.org>
```

File: tests/Unit/Libraries/Cms/Console/ExtensionDiscoverListCommandTest.php [hunk 7]
```diff
@@ -0,0 +1,88 @@
+<?php
+/**
+ * @package     Joomla.UnitTest
+ * @subpackage  Console
+ *
+ * @copyright   (C) 2019 Open Source Matters, Inc. <https://www.joomla.org>
```
Comment (PhilETaylor) [inline tests/Unit/Libraries/Cms/Console/ExtensionDiscoverListCommandTest.php]:
```suggestion
 * @copyright   (C) 2021 Open Source Matters, Inc. <https://www.joomla.org>
```

File: libraries/src/Console/ExtensionDiscoverCommand.php [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+	 *
+	 * @return  string  The text for the note
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	public function getNote(int $count): string
+	{
+		if ($count < 1)
+		{
+			return 'There is no extension to discover.';
```
Comment (PhilETaylor) [inline libraries/src/Console/ExtensionDiscoverCommand.php]:
```suggestion
			return 'No extensions were discovered.';
```

**Comment by PhilETaylor** [comment]
The code works perfectly, and has been tested, just made comments on text strings that can be improved and the file header dates - with those changed, this is ready :) 

Thanks!

**Comment by ceford** [comment]
I have tested this item :white_check_mark: successfully on be63829fcf441152441a9aa62fad671f416b90a1<br /><br />Commands worked as described - discovered and then installed one extension.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/33749">issues.joomla.org/tracker/joomla-cms/33749</a>.</sub>

**Comment by ghost** [comment]
@PhilETaylor I committed all your suggestions. Please have look, if everything is fine.

**Comment by PhilETaylor** [comment]
I have tested this item :white_check_mark: successfully on 497b50fa0e5d0eb7431c7d59b990925ec3eccf13<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/33749">issues.joomla.org/tracker/joomla-cms/33749</a>.</sub>

**Comment by alikon** [comment]
please fix https://ci.joomla.org/joomla/joomla-cms/43524/1/8
:smiley:

**Comment by PhilETaylor** [comment]
easy fix @agi-code in the unit test.

**Comment by alikon** [comment]
funny when we have a pr with unit/system code too :smile:

**Comment by PhilETaylor** [comment]
> funny when we have a pr with unit/system code too 😄

Don't get excited, the unit tests don't actually cover the functionality of the command, which would be very hard to do end-to-end testing on :-) They just test some text strings.

**Comment by alikon** [comment]
that's another issue
let me be happy to see a pr with system tests ..... whatever  

plus a cli pr :+1:

**Comment by PhilETaylor** [comment]
If I were being pedantic I would also remove the word "successfully"...

from 

```
 ! [NOTE] 4 extensions have been discovered successfully.   
```

They have been "discovered", that is enough to say 

```
 ! [NOTE] 4 extensions have been discovered.
```

**Comment by alikon** [comment]
I have tested this item :white_check_mark: successfully on 646a8893d4d88b235b5ed16d0b813793f75eda42<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/33749">issues.joomla.org/tracker/joomla-cms/33749</a>.</sub>

**Comment by alikon** [comment]
@PhilETaylor can you please test this  1

**Comment by PhilETaylor** [comment]
Cough cough https://github.com/joomla/joomla-cms/pull/33749#issuecomment-838884666

**Comment by richard67** [comment]
@PhilETaylor Well there were commits after your test ... but I've just checked, it was just texts .. so I restore your result in the issue tracker.

**Comment by richard67** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/33749">issues.joomla.org/tracker/joomla-cms/33749</a>.</sub>

**Comment by richard67** [comment]
Just texts I meant.

**Comment by PhilETaylor** [comment]
yes it was just tests and the unit tests now pass - that is good enough for me :)

**Comment by richard67** [comment]
> yes it was just tests and the unit tests now pass - that is good enough for me :)

@PhilETaylor Was a typo, not "tests", "texts" I meant. But it's ok.

**Pull request merged**